layer/input_buffer.cc

```diff
@@ -0,0 +1,209 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "input_buffer.h"
+
+#include <cassert>
+#include <cstdio>
+#include <vector>
+
+#include "absl/strings/str_cat.h"
+#include "logging.h"
+
+#if defined(__unix__)
+#include <fcntl.h>
+#include <sys/io.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#endif
+
+namespace performancelayers {
+
+namespace {
+// Cross-platfrom input buffer implementation that reads file contents using
+// fopen/fread/fclose.
+class FileInputBufferImpl : public InputBuffer::InputBufferImplBase {
+ public:
+  static absl::StatusOr<std::unique_ptr<InputBuffer::InputBufferImplBase>>
+  Create(const std::string& path) {
+    struct FileCloser {
+      void operator()(FILE* file) {
+        if (file) fclose(file);
+      }
+    };
+    std::unique_ptr<FILE, FileCloser> handle(fopen(path.c_str(), "rb"));
+    if (!handle) {
+      return absl::UnavailableError(
+          absl::StrCat("Failed to fopen file for read: ", path));
+    }
+
+    fseek(handle.get(), 0, SEEK_END);
```

Comment:
I do not have experiences using `fseek`, `fopen`, `fclose`, ... on Windows. Does it work fine?

Reply:
These come from `<cstdio>`. All standard library functions should work on all supported platforms.